File: problems/643/643.maximum-average-subarray-i.py
#
# @lc app=leetcode id=643 lang=python3
#
# [643] Maximum Average Subarray I
#


# @lc code=start
class Solution:
    def findMaxAverage(self, nums: List[int], k: int) -> float:
        curr_sum = sum(nums[:k])
        curr_max = curr_sum / k

        for r in range(k, len(nums)):
            curr_sum -= nums[r - k]
            curr_sum += nums[r]
            curr_max = max(curr_max, curr_sum / k)

        return curr_max


# @lc code=end

File: problems/643/643.maximum-average-subarray-i.java
/*
 * @lc app=leetcode id=643 lang=java
 *
 * [643] Maximum Average Subarray I
 */

// @lc code=start
class Solution {
    public double findMaxAverage(int[] nums, int k) {
        /*
         * find max average subarray length k in nums
         * k <= n
         *
         * intuition:
         * - sliding window
         * - keep track of max sum subarray
         *
         * O(N) time, O(1) space
         */

        double currMax = Integer.MIN_VALUE;
        int l = 0;
        double currSum = 0;

        for (int i = 0; i < nums.length; i++) {
            currSum += nums[i];

            if (i >= k - 1) {
                if (currSum / k > currMax)
                    currMax = currSum / k;

                currSum -= nums[l];
                l++;
            }
        }

        return currMax;
    }
}
// @lc code=end
